### ISSUE TITLE
Create `Problem` class to support generic `tf.keras.Model`

### ISSUE THREAD

**Issue by tomoleary** [tomoleary/hessianlearn#11]
The main issue is the loss function will have to be taken from the `model.losses` list multiplied by the associated weights, and this will also require that the data iterator construct can handle multiple inputs and outputs as inferred from the model class.

**Comment by tomoleary** [comment]
Validation should default to taking place at each "epoch equivalent" (i.e. `sweeps == 1`). At this point validation should iterate through the metrics of the model.

**Comment by tomoleary** [comment]
Perhaps an entirely new `HessianlearnModel` class needs to be written and the old one can be renamed something else. This way everything can inherit from `keras.Model` losses and metrics in a more straightforward / automatic way with less specific case checking based on `Problem` attributes.

**Comment by tomoleary** [comment]
> Perhaps an entirely new `HessianlearnModel` class needs to be written and the old one can be renamed something else. This way everything can inherit from `keras.Model` losses and metrics in a more straightforward / automatic way with less specific case checking based on `Problem` attributes.

If this is done, then the dipnet_paper examples in `hippyflow` will need to instance the old version of the HessianlearnModel class.

**Comment by tomoleary** [comment]
> > Perhaps an entirely new `HessianlearnModel` class needs to be written and the old one can be renamed something else. This way everything can inherit from `keras.Model` losses and metrics in a more straightforward / automatic way with less specific case checking based on `Problem` attributes.
> 
> If this is done, then the dipnet_paper examples in `hippyflow` will need to instance the old version of the HessianlearnModel class.

Actually what needs to happen is that a second `HessianlearnModel` class needs to be written to support things directly inheriting from keras, while the old one still is in existence to support the old way of doing things concurrently.

**Comment by tomoleary** [comment]
Multiple losses can be passed into the `compile` method, when the network has multiple outputs. In this case loss weights can also be passed in as a list of the same length. These are ordered the same as the outputs passed as lists in the constructor. When there is just one output there is only one loss; so calling the `add_loss` method will just sum the losses in place.

**Comment by tomoleary** [comment]
Issues so far: Using the losses and metrics from `keras.Model.loss_functions` and `keras.Model.metrics` I get the following issues

- When calling `keras.Model.evaluate` in advance of constructing the hessianlearn model I get: `Resource localhost/total/N10tensorflow3VarE does not exist.`  [This seems to be a v1.x issue](https://stackoverflow.com/questions/60285770/resource-localhost-total-n10tensorflow3vare-does-not-exist)
- When calling `keras.Model.evaluate` after the hessianlearn `fit` I get `You must feed a value for placeholder tensor 'output_placeholder' with dtype float and shape [?,OUTPUTSHAPE]` when clearly this is being passed in correctly.

**Comment by tomoleary** [comment]
> Issues so far: Using the losses and metrics from `keras.Model.loss_functions` and `keras.Model.metrics` I get the following issues
> 
> * When calling `keras.Model.evaluate` in advance of constructing the hessianlearn model I get: `Resource localhost/total/N10tensorflow3VarE does not exist.`  [This seems to be a v1.x issue](https://stackoverflow.com/questions/60285770/resource-localhost-total-n10tensorflow3vare-does-not-exist)
> * When calling `keras.Model.evaluate` after the hessianlearn `fit` I get `You must feed a value for placeholder tensor 'output_placeholder' with dtype float and shape [?,OUTPUTSHAPE]` when clearly this is being passed in correctly.

[This seems to be somehow a ramification of session ambiguity.](https://github.com/tensorflow/tensorflow/issues/38571)

**Comment by tomoleary** [comment]
> > Issues so far: Using the losses and metrics from `keras.Model.loss_functions` and `keras.Model.metrics` I get the following issues
> > 
> > * When calling `keras.Model.evaluate` in advance of constructing the hessianlearn model I get: `Resource localhost/total/N10tensorflow3VarE does not exist.`  [This seems to be a v1.x issue](https://stackoverflow.com/questions/60285770/resource-localhost-total-n10tensorflow3vare-does-not-exist)
> > * When calling `keras.Model.evaluate` after the hessianlearn `fit` I get `You must feed a value for placeholder tensor 'output_placeholder' with dtype float and shape [?,OUTPUTSHAPE]` when clearly this is being passed in correctly.
> 
> [This seems to be somehow a ramification of session ambiguity.](https://github.com/tensorflow/tensorflow/issues/38571)

I can get around this issue clunkily as it may be by calling `tf.keras.Model.compile` in between interfacing with hessianlearn and calling `tf.keras.Model.evaluate`

**Comment by tomoleary** [comment]
The new issue is about the metrics returning rolling averages. Everytime the metrics are called directly in reference to the item in `tf.keras.Model.metrics`, [I should probably make sure that `tf.keras.Model.reset_states()` is also called first](https://keras.io/guides/customizing_what_happens_in_fit/).

**Comment by tomoleary** [comment]
```
loss_things =  ['_callable_losses', '_clear_losses', '_compile_weights_loss_and_weighted_metrics', '_eager_losses', '_feed_loss_fns', '_graph_network_add_loss', '_loss_weights_list', '_losses', '_output_loss_metrics', '_prepare_total_loss', '_recompile_weights_loss_and_weighted_metrics', 'add_loss', 'get_losses_for', 'loss', 'loss_functions', 'loss_weights', 'losses', 'total_loss']
metric_things =  ['_add_unique_metric_name', '_cache_output_metric_attributes', '_compile_metric_functions', '_compile_metrics', '_compile_weighted_metrics', '_eager_add_metric', '_get_existing_metric', '_get_training_eval_metrics', '_graph_network_add_metric', '_handle_metrics', '_handle_per_output_metrics', '_init_metric_attributes', '_metrics', '_per_output_metrics', '_per_output_weighted_metrics', '_set_metric_attributes', '_set_per_output_metric_attributes', '_symbolic_add_metric', 'add_metric', 'metrics', 'metrics_names', 'reset_metrics']
```

**Comment by tomoleary** [comment]
Allow hessianlearn model to take no regularization, then just default it to `L2Regularization` with `gamma = 0`

**Comment by tomoleary** [comment]
Additionally `dtype` can be inferred from `tf.keras.Model.inputs[0].dtype` but perhaps this could cause a problem somewhere?

**Comment by tomoleary** [comment]
Need to update the `Data` object to be able to handle multiple inputs and multiple outputs

**Comment by tomoleary** [comment]
> Need to update the `Data` object to be able to handle multiple inputs and multiple outputs

Actually this seems to not be an issue, since it handles generic dictionaries key'd on whatever. So this must just be handled at the level of querying the `tf.Session` which will happen inside of the modelWrapper

**Comment by tomoleary** [comment]
Different losses for different outputs:

https://stackoverflow.com/questions/49583805/using-different-loss-functions-for-different-outputs-simultaneously-keras
